src/dmd/dtemplate.d

```diff
@@ -8306,28 +8310,89 @@ struct TemplateStats
     /*******************************
      * Add this unique instance
      */
-    static void incUnique(const TemplateDeclaration td)
+    static void incUnique(const TemplateDeclaration td,
+                          const TemplateInstance ti)
     {
         if (!global.params.vtemplates)
             return;
         if (!td)
             return;
+        assert(ti);
         if (auto ts = cast(const void*) td in stats)
+        {
+            if (ts.uniqueInstances is null)
+                ts.uniqueInstances = new TemplateInstances();
+            if (global.params.vtemplatesListUnique)
+                ts.uniqueInstances.push(cast() ti);
             ++ts.uniqueInstantiations;
+        }
         else
             stats[cast(const void*) td] = TemplateStats(0, 1);
     }
 }
 
-
 void printTemplateStats()
 {
+    static struct TemplateDeclarationStats
+    {
+        TemplateDeclaration td;
+        TemplateStats ts;
+        static int compare(scope const TemplateDeclarationStats* a,
+                           scope const TemplateDeclarationStats* b) @safe nothrow @nogc pure
+        {
+            return b.ts.uniqueInstantiations - a.ts.uniqueInstantiations;
+        }
+    }
+
     if (!global.params.vtemplates)
         return;
 
-    printf("  Number   Unique   Name\n");
-    foreach (td, ref ts; TemplateStats.stats)
+    Array!(TemplateDeclarationStats) sortedStats;
+    sortedStats.reserve(TemplateStats.stats.length);
+    foreach (td_, ref ts; TemplateStats.stats)
+    {
+        sortedStats.push(TemplateDeclarationStats(cast(TemplateDeclaration) td_, ts));
+    }
+
+    sortedStats.sort!(TemplateDeclarationStats.compare);
+
+    foreach (const ref ss; sortedStats[])
     {
-        printf("%8u %8u   %s\n", ts.numInstantiations, ts.uniqueInstantiations, (cast(const TemplateDeclaration) td).toChars());
+        if (global.params.vtemplatesListUnique &&
+            ss.ts.uniqueInstances)
+        {
+            message(ss.td.loc,
+                    "vtemplate: found %u (%u unique) instantiation(s) of template `%s`, unique instances are:",
+                    ss.ts.numInstantiations,
+                    ss.ts.uniqueInstantiations,
+                    ss.td.toChars());
+            foreach (const ti; (*ss.ts.uniqueInstances)[])
+            {
+                TemplateInstance tinst = cast()ti.tinst;
+
+                if (tinst)
+                    message(ti.loc, "vtemplate: bottom instance `%s`, enclosing instance is:", ti.toChars());
+                else
+                    message(ti.loc, "vtemplate: top instance `%s`", ti.toChars());
+
+                while (tinst)
```

Comment:
This _cannot_ trace upwards through `alias` and `enum` definitions such as

enum bool isIntegral(T) = is(IntegralTypeOf!T) && !isAggregateType!T;

but stops at `IntegralTypeOf`. This is a severe limitation I would like to fix.

Which dmd-classes represent `alias` and `enum` with template params? I find 3 definitions of `AliasDeclaration`. Which one should I investigate?

Any clues on how to solve this?

Reply:
Ping, @MoonlightSentinel.

Reply:
Sorry, i don't know. Maybe you can adapt the Visitor methods for those nodes to propagate `tinst` properly?